File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []
        curr_res = []
        def dfs():
            if len(curr_res) == len(nums):
                res.append(curr_res[:])#append copy of curr_res, cuz list is mutable
                return
            #recursive case:
            for i in nums:
                if i not in curr_res:
                    curr_res.append(i)
                    dfs() 
                    curr_res.pop()#so that we go back to the for loop again, do another dfs
        dfs()
        return res
            

File: 0046-permutations/0046-permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> currRes = new ArrayList<>();
        
        dfs(nums, res, currRes);
        return res;
    }
    
    public void dfs(int[] nums, List<List<Integer>> res, List<Integer> currRes) {
        if (currRes.size() == nums.length) {
            res.add(new ArrayList<>(currRes));
            return;
        }
        
        for (int n : nums) {
            if (!currRes.contains(n)) {
                currRes.add(n);
                dfs(nums, res, currRes);
                currRes.remove(currRes.size() - 1);
            }
        }
    }
}